apps/showcase/src/components/training/code-editor-view/code-editor-view.component.ts

```diff
@@ -316,10 +339,48 @@ export class CodeEditorViewComponent implements OnDestroy {
         }
       });
     });
+    void this.retrieveSaveChanges();
+  }
+
+  private async retrieveSaveChanges() {
+    await firstValueFrom(this.cwdTree$);
+    const { cwd } = this.project();
+    const savedState = localStorage.getItem(cwd);
+    if (!savedState) {
+      return;
+    }
+    const state = JSON.parse(savedState);
+    const hasDiscrepancies = (await Promise.all(
+      Object.entries(state).map(async ([path, text]) =>
+        text !== (await this.webContainerService.readFile(`${this.project().cwd}/${path}`).catch(() => undefined))
+      ))
+    ).some((hasDiscrepancy) => !!hasDiscrepancy);
```

Comment:
We don't have to wait the response for all the files to dhow the dialog, as soon as we spot one discrepancy
Maybe something like:
```suggestion
    const hasDiscrepancies = (await Promise.any(
      Object.entries(state).map(([path, text]) => new Promise((resolve, reject) =>
        text !== (await this.webContainerService.readFile(`${this.project().cwd}/${path}`).catch(() => undefined)) ? resolve(true) : reject()
      ))
    ).catch(() => false);
```